### DISCUSSION TITLE
How to detect events on each segment of a Polyline

### DISCUSSION THREAD

**Discussion by Pacobzt** [peacefulotter/react-leaflet-hotline#5] [General]
Hi, I am using react-leaflet-hotline to display drifting buoy trajectories. I need to display multiple Polylines, and to detect events on each segment of each Polyline. It works on the first render of the Hotline component, but when I update my data it no longer works and I can't find a fix.

Here is my simplified code:

- I have those datasets
```
const data1 = [
  [
    [
      { lat: 50.69565, lng: -2.76608, value: 0.28 },
      { lat: 50.69376, lng: -2.76531, value: 0.27 },
    ],
    [
      { lat: 50.69376, lng: -2.76531, value: 0.27 },
      { lat: 50.68521, lng: -2.76671, value: 0.29 },
    ],
  ],
];
const data2 = [
  [
    [
      { lat: 50.68521, lng: -2.76671, value: 0.29 },
      { lat: 50.67754, lng: -2.77558, value: 0.31 },
    ],
    [
      { lat: 50.67754, lng: -2.77558, value: 0.31 },
      { lat: 50.67112, lng: -2.78789, value: 0.37 },
    ],
  ],
  [
    [
      { lat: 50.67112, lng: -2.78789, value: 0.37 },
      { lat: 50.66361, lng: -2.81011, value: 0.34 },
    ],
  ],
];
```

- I have this defined in my react context
`const [segments, setSegments] = useState(data1);`

- Then I use it in my Hotline component
```
import { useMemo } from "react";
import { Hotline } from "react-leaflet-hotline";
import { useMap } from "src/contexts/values/UseMap";

export default function Trajectories() {
  const { segments } = useMap();

  const eventHandlers = useMemo(
    () => ({
      click: (e, i, p) => console.log(i),
      mouseover: (e, i, p) => p.setStyle({ opacity: 0.5 }),
      mouseout: (e, i, p) => p.setStyle({ opacity: 0 }),
    }),
    []
  );

  return segments.map((buoyGroup, idx) => (
    <Hotline
      key={idx}
      data={buoyGroup}
      getLat={(t: any) => t.lat}
      getLng={(t: any) => t.lng}
      getVal={(t: any) => t.value}
      eventHandlers={eventHandlers}
    />
  ));
}
```

- Here is how I update data
```
export default function Commands() {
  const { setSegments } = useMap();
  return (
    <>
      <Button onClick={() => setSegments(data1)}>2 segment</Button>
      <Button onClick={() => setSegments(data2)}>3 segment</Button>
      <Button onClick={() => setSegments([])}>clear</Button>
    </>
  );
}
```

The events work correctly on the first render of my Hotline component (with "data1"), and if I update the data it displays correctly but the events no longer work.

Thank you for your answer,
Paco

**Comment by peacefulotter** [comment]
(Answering on the phone, will make it more clear later) 

Thank you for taking the time to make a minimal repro. I am not sure what causes the issue and I can't test your repro before this weekend. 

In the meantime, I can see that you have multiple duplicate points, e.g. [(a, b), (b, c), (c, d)]; could you try to flatten the data arrays such that it is an array of coordinates e.g. [a, b, c, d] ? That way you only need a single Hotline component. 

I think the issue is multiple hotline components create multiple canvas which messes up the rendering system.

**Reply by Pacobzt** [reply]
I managed to get the code to work by refreshing the react-leaflet "MapContainer" between each data update instead of the Hotline.

Thanks for your comment on flattening the data. I already tried but events are detected on the whole polyline if I do this, and I want to detect events on each segment of my polyline. Maybe there is still a cleaner way to do this?...

**Reply by peacefulotter** [reply]
Ok great, sorry for the troubles. I will try to make this work properly this week-end.
If you are allowed to share some screenshots, it would be a pleasure to see the end result!

**Reply by peacefulotter** [reply]
Found the issue, it was actually a big deal : the polyline would only update if new segments are added. Additionally, it made me find a typing issue where getLat, getLng and getVal would have the wrong callback type on 2D coordinates arrays. Both fixes are coming tomorrow.

You will be able to do the following, with events on each individual line: 
```tsx
'use client'

import { useMemo, useState } from 'react'
import { MapContainer, TileLayer } from 'react-leaflet'
import { Hotline } from 'react-leaflet-hotline'
import { HotlineEventHandlers } from 'react-leaflet-hotline/dist/types/types'

import 'leaflet/dist/leaflet.css'

const data1 = [
  [
    { lat: 50.69565, lng: -2.76608, value: 0.28 },
    { lat: 50.69376, lng: -2.76531, value: 0.27 },
  ],
  [
    { lat: 50.69376, lng: -2.76531, value: 0.27 },
    { lat: 50.68521, lng: -2.76671, value: 0.29 },
  ],
]

const data2 = [
  [
    { lat: 50.68521, lng: -2.76671, value: 0.29 },
    { lat: 50.67754, lng: -2.77558, value: 0.31 },
  ],
  [
    { lat: 50.67754, lng: -2.77558, value: 0.31 },
    { lat: 50.67112, lng: -2.78789, value: 0.37 },
  ],
  [
    { lat: 50.67112, lng: -2.78789, value: 0.37 },
    { lat: 50.66361, lng: -2.81011, value: 0.34 },
  ],
]

type Data = { lat: number; lng: number; value: number }

export default function Map() {
  const [segments, setSegments] = useState<Data[][]>([])

  const eventHandlers: HotlineEventHandlers = useMemo(
    () => ({
      click: console.log,
    }),
    []
  )

  return (
    <div>
      <div>
        <button onClick={() => setSegments(data1)}>2 segment</button>
        <button onClick={() => setSegments(data2)}>3 segment</button>
        <button onClick={() => setSegments([])}>clear</button>
      </div>
      <MapContainer
        center={[50.68521, -2.76671]}
        zoom={12}
        scrollWheelZoom={true}
        style={{ height: '80vh', width: '80vw' }}
      >
        <TileLayer ... />
        <Hotline
          data={segments}
          getLat={({point}) => point.lat}
          getLng={({point}) => point.lng}
          getVal={({point}) => point.value}
          eventHandlers={eventHandlers}
        />
      </MapContainer>
    </div>
  )
}

```

**Reply by peacefulotter** [reply]
Alright, fixes published under 1.6.1 ! Couple of notes:
- I defined `eventHandlers` as type HotlineEventHandlers for type safety and autocompletion
- The API for Hotline component changed. Specifically the getters (getLat / getLng, getVal) now pass an object as callback parameter. Also, you now have both segment index and line index accessible via the getters. Basically, a segment is a collection of lines (e.g. for your case where you have multiple segments)
- Same applies for the eventHandlers callbacks, passing an object (but without the line index since this is not currently supported)

I still need to mention some of these changes somewhere in the docs, sorry that its not so clear right now

**Reply by Pacobzt** [reply]
Thank you very much for your answers and your work. This is how I use your package:
![Screenshot from 2024-12-17 13-30-06](https://github.com/user-attachments/assets/61478c46-74f5-45c5-a5eb-a2d30cf51efa)

**Reply by peacefulotter** [reply]
Amazing! Thank you so much for using my package 😀 
Do not hesitate to reach out if you need anything
